BUG: Fix warning problems of the mod operator

arubiales opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

That's the initial solution, there are 13 test failing because it does not trigger the warning that we are resolving. So If this solution is correct I think we should change the test too.

arubiales committed:
solve the problems of mod operator except for infinite values #18170

seberg commented:
Could you fix up the tests (and maybe even add new ones to describe what you changed?).  That also clarifies exactly what changed for the reviewers.

arubiales committed:
Solved test fails due to the warnings changed #18170

arubiales commented:
@seberg  did it! Check the changes when you can please.

arubiales committed:
Fix 1 test_umath and npy_math_internal

arubiales commented:
@seberg did it! Check the changes when you can please.

nschloe commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -716,9 +710,6 @@ npy_divmod@c@(@type@ a, @type@ b, @type@ *modulus)
     @type@ div, mod, floordiv;
 
     /* force set invalid flag, doesnt raise by default on gcc < 8 */
```
This line should probably go, too.

seberg replied:
This is my "problem" so to speak.  I am confused what this is about and what we tried to do in gh-16161.

My guess is that [this godbolt snippet explains the comment](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:27,positionColumn:2,positionLineNumber:27,selectionStartColumn:2,selectionStartLineNumber:27,startColumn:2,startLineNumber:27),source:'//+Type+your+code+here,+or+load+an+example.%0A%23include+%3Cmath.h%3E%0A%23include+%3Cfenv.h%3E%0A%23include+%3Cstdio.h%3E%0A%0Aint%0Atest(double+num)%0A%7B%0A++++if+(num+%3E+0)+%7B%0A++++++++return+1%3B%0A++++%7D%0A++++return+0%3B%0A%7D%0A%0A%0A%0Aint%0Amain(void)%0A%7B%0A++++double+value+%3D+NAN%3B%0A++++feclearexcept(FE_ALL_EXCEPT)%3B%0A%0A++++int+res+%3D+test(value)%3B%0A++++int+errs+%3D+fetestexcept(FE_ALL_EXCEPT)%3B%0A++++printf(%22Should+not+be+0,+should+have+an+exception+set:+%25d%5Cn%22,+errs)%3B%0A++++return+res%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:g73,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'0',trim:'1'),fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'x86-64+gcc+7.3+(Editor+%231,+Compiler+%231)+C%2B%2B',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+7.3+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4) if you switch to GCC 8, the correct answer is given (warning flag is set). 

Now... If `a is NaN` and `b < 0`, the `if (mod)` will evaluate `True` if mod is NaN. Especially, if both `a` and `b` are `NaN`, then we definitely take that branch and both `b < 0` and `mod < 0` evaluate to False **with** setting the warning (since they are both NaN).  And that was *wrong* on GCC 7, which did not set the warning (but led to the *better* result).

(Hopefully, that flow analysis is right... But something is off and my best guess is to fix most of these comparisons and replace them with `isgreater` and `isless`, as mentioned before and now also in gh-19398.  In general, I don't think we really ever want float comparisons to set warning flags really, so de-facto *all* of those in the math code are probably just wrong...)

I honestly don't understand why:
```
with np.errstate(invalid='raise'):
    assert_no_warnings(FloatingPointError, np.floor_divide, 1, np.nan)
```
succeeds!  If I just call `np.floor_divide(1, np.nan)`, I get an "invalid value" warning, and with `np.errstate(invalid="raise")` the `FloatingPointError`.

seberg replied:
@arubiales I am confused, did you maybe try these things out on a gcc version older than 8 (Which would be pretty old)?

arubiales replied:
@nschloe Yes sorry I fogot to delete it.  

@seberg No my GCC version is 8.4.  I try the same, and with the changes I have no warning, but I'm with you, should raise the "invalid value" flag. After test with this change:
```
NPY_INPLACE @type@
npy_floor_divide@c@(@type@ a, @type@ b) {
    @type@ div, mod;
    if (NPY_UNLIKELY(!b)) {
        div = a / b;
        if (!a || npy_isnan(a)) {
            npy_set_floatstatus_invalid();
        } else {
            npy_set_floatstatus_divbyzero();
        }
    } else {
        div = npy_divmod@c@(a, b, &mod);
        /* This is the change, I just set the invalid float status in the else */
        npy_set_floatstatus_invalid();
    }
```

 Which mean I never enter in the `if` because `NPY_UNLIKELY(!b)` is **always** `False`

seberg replied:
Yeah, `!b` is false, but then it goes into divmod?  I have to check more, maybe I was also missing a `git clean`...

Can you do me a favor and delete the `assert_no_warnings` completely and make it `np.errstate(all="raise")`?  I still have to think about it more.  I guess this is a good change though, and we should have more fixes as followup.

nschloe commented:
Perhaps and `assert_no_warnings` test for `np.mod` could be added.

arubiales committed:
delete comment about warning

arubiales commented:
@nschloe Do you want to add it in this PR? A new test function as for example `divmod` but for `mod`?

seberg committed:
TST: Change tests to more strictly (correctly) check for errors

The old tests seems not to have noticed all types of errors here.
It seems the `assert_no_warnings` context does not care about
exceptions.

seberg commented:
I have pushed a test commit to remove `assert_no_warnings`.  Unfortunately, that should blow up the tests (if it does not, I am very suprised, since they fail locally).

I am thinking about how to fix things, but I am not sure how easy it will be to confine changes to a minimal set here.

seberg committed:
BUG: Fix floating point error flags in division related ops

seberg commented:
Sorry, for taking over a bit here, but I had to look at it with much more concentration to make progress.  It is a bit tricky to split the `mod` issue from the other issues :/.

@charris maybe you can have a look? I think this should be fine now, although a couple of new tests might be good (Although Anirudh had added good tests and the only problem is the behaviour with NaNs coming in).

The things to remember is:
* We raise errors whenever a **new** `inf` or `NaN` is created.  If a new `inf` is created, it will be a division-by-zero, and if a new NaN is created it will be an "invalid value".
* `if NaN` takes the True branch (evaluates True); `if !NaN:` evaluates to false.
* I assume `divmod` can create _both_ a new NaN and a new Inf, so sets both flags.
* floor-divide and remainder only set one of those.  When `divmod` would set both flags we cannot use it.
* `NaN > 0` sets an invalid value flag.  We need to avoid that (see gh-19398), using `isgreater`, etc. should not set the flags.

The last point is a big reason for the whole confusion here:
* Old GCC versions (or some more?) did not set a warning flag for `NaN` comparisons (note that `==` and `!=` do not set the warning)
* We also use Python's divmod, but Python does not really care about floating point exceptions to begin with.

`isgreater`, etc. are C99 standard, so I assume we can use them.

seberg commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -412,32 +412,6 @@ NPY_INPLACE @type@ npy_@kind@@c@(@type@ x, @type@ y)
-}
-#endif
-/**end repeat1**/
-
```
These `fmod` implementations were added in the old divmod-warnings PR.  Removing them undoes that change.

seberg commented on numpy/core/src/npymath/npy_math_internal.h.src:
```diff
@@ -681,28 +632,33 @@ NPY_INPLACE @type@
+         * but set the error flags correct (divmod may set too many).
+         * (Using !b here, since reverse logic set incorrect flags on clang)
+         */
+        return npy_fmod@c@(a, b);
```
What I had missed before is that this also fails the following test for `divmod`: https://github.com/numpy/numpy/blob/6496f1d6abf83cdad37ecdac4453c432ff2f8573/numpy/core/tests/test_umath.py#L579-L580

`fmod(1, 0)` seems to simply not set the correct floating point flags on clang (at least in the version used by CI).  I suppose we could work around clang (or even mark the test as xfail?)

arubiales commented:
> Sorry, for taking over a bit here

Don't worry, you have more knowledge that me about Numpy. I'm following all your changes, if there is something I can help, please tell me, I will be happy to help!

seberg commented:
Thanks.  Right now I am thinking of calling it done.  The tests will fail, and I will fix them up so that it skips the shaky ones at least on MacOS.  (With a good chance of more skips needed for weird platforms.)

I have two options:
1. My slight preference for now:  Regress for `%`/`divmod` (`np.floor_divide` seems fine) to the behaviour in gh-14900 for MacOS.  This is due to `fmod` not setting the correct warning flags on MacOS.  This partially goes back to the behaviour of 1.19.x (and earlier).  (MacOS used to not give any warnings for `np.divmod(1., 0.)`)  But fixes a bunch of other things, and especially makes sure no spurious warnings are given. – If someone wants MacOS to be fixed, option 2 can be done later.
2. Manually implement `fmod` (potentially selective for MacOS using the blocklist) that implements setting the floating point flags manually.

I dislike testing for correct flags a bit.  It feels like this is bound to fail on exotic platforms.  I am honestly wondering if rather than failing, these tests should just print out: "Careful, your system does not follow IEEE warning flags well for divmod"

seberg committed:
TST: Ignore floating point warning tests on MacOS

It is very much possible that this will also fail on other systems
and should rather only run, e.g. on newer glibc versions.

In principle, we could also split this into "no warnings given"
and "warning given".  MacOS seems very sloppy with floating point
warnings, but hopefully it at least does not give spurious ones.

mattip commented:
We discussed this at the triage meeting and decided to merge this as is. In a follow on PR we should mention that macOS (probably)/clang (less likely) is not IEEE compliant.

mattip commented:
Thanks @seberg
